### PR TITLE
Faster URIUtils::HasExtension Checks

### PR DESCRIPTION
**Pull request by CrystalP**
## Description

Replacement of  #23729 and related discussion in #23726

URIUtils::HasExtension works with lists of concatenated extensions separated by | (ex. ".mkv|.mp4|.avi|.m4v")
PR above wanted to improve performance by splitting the extensions before hand and storing them as vectors of strings, ex. { ".mkv", ".mp4", ".avi", ".m4v"}

Testing showed (see benchmarks in testing) that improved parsing of the current format of extensions lists provides performance gains similar to the use of vectors of strings, without having to change all the callers. Two parsing-improving algorithms are included in the PR.

To improve performance beyond that, std::unordered_set is needed (using hashing) but requires changes to lots of callers, all over the code base.

The current algorithm spends a lot of time lowercasing then splitting the list into a vector of strings, the new algorithms parse the string and process piece by piece.

* Two improved algorithms:
  - commit 2, straight forward, readable and C++like (2x to 20x faster than current)
  - commit 3, less readable but even faster (4x to 68x faster than current and comparable to the use of a vector of strings).
The speedup for both depends largely on the position of the matching extension in the string of extensions.

* Added many unit tests to prove the algorithms provide the same results as master, quirks included :-)

* Added a cache to the GetxxxExtensions() family of functions (ex GetVideoExtensions), which otherwise build strings to return for every call, concatenating some strings. They only need to do the concatenation once on profile load or addon addition/removal.

* Minor ~5% perf. improvement of StringUtils::EqualsNoCase / StringUtils::EndsWithNoCase if they remain (ie algo of commit 3 is chosen)

It would be possible to keep the string_view overloads of StringUtils functions added for commit 2 instead of cleaning them up.

Future: reordering the extensions in the builtin lists would provide a nice speedup as well and would allow achieving the higher range of the possible improvements for popular extensions, rather than the lower range.

copied @garbear @ksooo @howie-f as you commented on the original PR.

## Motivation and context

Noticed HasExtension() function appear in a flame graph of a library scan of movies and #23729 meant to address a similar problem with artwork.

## How has this been tested?

Added unit tests covering every path in the new algorithms and clarifying corner cases of the existing code.
Kodi runs fine!

### Benchmark results

short list of 4 extensions - ".mkv|.mp4|.avi|.m4v"
| algorithm / type | debug build | release build | comment |
|:---|:---|:---|:---|
| Current master | 1x | 1x | |
| Improved algorithm 1 (commit 2) | 2x-4x | 2.5x-3.2x | depends on position of matched extension |
| Improved algorithm 2 (commit 3) | 8x | 4x-6x | depends on position of matched extension |
| vector/array of strings (not in PR) | 2.2x-3.5x | 3.x-4.3x | depends on position of matched extension |
| unordered_set (not in PR) | 9x-10x | 6.3x-6.7x | hash collisions |

long list of ~90 extensions - GetFileExtensionProvider().GetVideoExtensions()
| algorithm / type | debug build | release build | comment |
|:---|:---|:---|:---|
| Current master | 1x | 1x | |
| Improved algorithm 1 (commit 2) | 2x-20x | 1.7x-15x | depends on position of matched extension |
| Improved algorithm 2 (commit 3) | 26x-68x | 3.5x-23x | depends on position of matched extension |
| vector/array of strings (not in PR) | 2.3x-29x | 2.5x-25x | depends on position of matched extension |
| unordered_set (not in PR) | 120x | 50x-58x | rather constant hash lookup time |

The cached versions of the GetxxxExtensions() family of functions (ex GetVideoExtensions) are 60% faster than the original for debug builds. I wasn't able to measure a difference in release builds, due to testing methodology maybe, or aggressive compiler inlining/optimization...

## What is the effect on users?

Overall small performance improvement, likely more noticeable on slow hardware in library scans.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [X] I have added tests to cover my change
- [X] All new and existing tests passed

### CODE REVIEW

File: xbmc/settings/AdvancedSettings.h [hunk 1]
```diff
@@ -100,6 +103,8 @@ struct RefreshVideoLatency
 
 typedef std::vector<TVShowRegexp> SETTINGS_TVSHOWLIST;
 
+typedef std::function<void()> AdvancedSettingsCallback;
```
Comment (ksooo) [inline xbmc/settings/AdvancedSettings.h:106]: We prefer `using`, not `typedef`.

=> https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#127-type-aliases

File: xbmc/settings/AdvancedSettings.h [hunk 2]
```diff
@@ -116,6 +121,9 @@ class CAdvancedSettings : public ISettingCallback, public ISettingsHandler
     void AddSettingsFile(const std::string &filename);
     bool Load(const CProfileManager &profileManager);
 
+    int RegisterSettingsLoadedCallback(AdvancedSettingsCallback callback);
+    void UnregisterSettingsLoadedCallback(int);
```
Comment (ksooo) [inline xbmc/settings/AdvancedSettings.h:125]: Have you intentionally not provided a name for the parameter?

File: xbmc/settings/AdvancedSettings.h [hunk 3]
```diff
@@ -386,4 +394,7 @@ class CAdvancedSettings : public ISettingCallback, public ISettingsHandler
     void Initialize();
     void Clear();
     void SetExtraArtwork(const TiXmlElement* arttypes, std::vector<std::string>& artworkMap);
+
+    mutable CCriticalSection m_critSection; //!< protects the callbacks list access
```
Comment (ksooo) [inline xbmc/settings/AdvancedSettings.h:398]: If this mutex is only for protecting the callbacks list, like the comment suggests, you probably should name the member accordingly, then the comment would even not be needed anymore.

File: xbmc/utils/FileExtensionProvider.cpp [hunk 4]
```diff
@@ -48,53 +52,77 @@ CFileExtensionProvider::~CFileExtensionProvider()
...
 }
 
 std::string CFileExtensionProvider::GetMusicExtensions() const
 {
-  std::string extensions(m_advancedSettings->m_musicExtensions);
-  extensions += '|' + GetAddonExtensions(AddonType::VFS);
-  extensions += '|' + GetAddonExtensions(AddonType::AUDIODECODER);
+  if (!m_musicExtensions)
+  {
+
```
Comment (ksooo) [inline xbmc/utils/FileExtensionProvider.cpp:65]: remove the blank line

File: xbmc/utils/FileExtensionProvider.cpp [hunk 5]
```diff
@@ -48,53 +52,77 @@ CFileExtensionProvider::~CFileExtensionProvider()
...
 }
 
 std::string CFileExtensionProvider::GetPictureExtensions() const
 {
-  std::string extensions(m_advancedSettings->m_pictureExtensions);
-  extensions += '|' + GetAddonExtensions(AddonType::VFS);
-  extensions += '|' + GetAddonExtensions(AddonType::IMAGEDECODER);
+  if (!m_pictureExtensions)
+  {
+
```
Comment (ksooo) [inline xbmc/utils/FileExtensionProvider.cpp:78]: remove the blank line

File: xbmc/utils/FileExtensionProvider.cpp [hunk 6]
```diff
@@ -207,6 +235,10 @@ void CFileExtensionProvider::SetAddonExtensions(AddonType type)
           fileFolderExtensions.push_back(ext.first);
       }
     }
+    // Invalidate dependent cached extensions lists
+    m_musicExtensions = {};
+    m_pictureExtensions = {};
+    m_fileFolderExtensions = {};
```
Comment (ksooo) [inline xbmc/utils/FileExtensionProvider.cpp:241]: `m_subtitlesExtensions` is missing here?!

Comment (ksooo) [reply xbmc/utils/FileExtensionProvider.cpp:241]: What about `m_discStubExtensions`?

File: xbmc/utils/FileExtensionProvider.cpp [hunk 7]
```diff
@@ -230,6 +262,11 @@ void CFileExtensionProvider::SetAddonExtensions(AddonType type)
         }
       }
     }
+    // Invalidate dependent cached extensions lists
+    m_musicExtensions = {};
+    m_pictureExtensions = {};
+    m_subtitlesExtensions = {};
+    m_videoExtensions = {};
```
Comment (ksooo) [inline xbmc/utils/FileExtensionProvider.cpp:269]: What about `m_discStubExtensions`?

File: xbmc/utils/FileExtensionProvider.h [hunk 8]
```diff
@@ -77,15 +78,25 @@ class CFileExtensionProvider
   void SetAddonExtensions(ADDON::AddonType type);
 
   void OnAddonEvent(const ADDON::AddonEvent& event);
+  void OnAdvancedSettingsLoaded();
 
   // Construction properties
   std::shared_ptr<CAdvancedSettings> m_advancedSettings;
   ADDON::CAddonMgr &m_addonManager;
+  int m_callbackId;
```
Comment (ksooo) [inline xbmc/utils/FileExtensionProvider.h:86]: default init the member

File: xbmc/utils/StringUtils.h [hunk 9]
```diff
@@ -104,9 +104,9 @@ class StringUtils
   static void ToLower(std::wstring &str);
   static void ToCapitalize(std::string &str);
   static void ToCapitalize(std::wstring &str);
-  static bool EqualsNoCase(const std::string &str1, const std::string &str2);
-  static bool EqualsNoCase(const std::string &str1, const char *s2);
-  static bool EqualsNoCase(const char *s1, const char *s2);
+  static bool EqualsNoCase(const std::string& str1, const std::string& str2);
+  static bool EqualsNoCase(const std::string& str1, const char* s2);
+  static bool EqualsNoCase(const char* s1, const char* s2);
```
Comment (ksooo) [inline xbmc/utils/StringUtils.h:109]: Why do you only fix these formatting issues, not all/none still contained in this header? Change seems completely random.

File: xbmc/utils/StringUtils.h [hunk 10]
```diff
@@ -141,8 +141,8 @@ class StringUtils
   static bool StartsWithNoCase(const char *s1, const char *s2);
   static bool EndsWith(const std::string &str1, const std::string &str2);
   static bool EndsWith(const std::string &str1, const char *s2);
-  static bool EndsWithNoCase(const std::string &str1, const std::string &str2);
-  static bool EndsWithNoCase(const std::string &str1, const char *s2);
+  static bool EndsWithNoCase(const std::string& str1, const std::string& str2);
+  static bool EndsWithNoCase(const std::string& str1, const char* s2);
```
Comment (ksooo) [inline xbmc/utils/StringUtils.h:145]: Why do you only fix these formatting issues, not all/none still contained in this header? Change seems completely random.

File: xbmc/utils/URIUtils.h [hunk 11]
```diff
@@ -67,8 +82,8 @@ class URIUtils
          strExtensions is empty.
    \sa GetExtension
    */
-  static bool HasExtension(const std::string& strFileName, const std::string& strExtensions);
-  static bool HasExtension(const CURL& url, const std::string& strExtensions);
+  static bool HasExtension(const std::string& strFileName, std::string_view strExtensions);
+  static bool HasExtension(const CURL& url, std::string_view strExtensions);
```
Comment (ksooo) [inline xbmc/utils/URIUtils.h:86]: Could you please remove the type prefix from the string params?

=> https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#87-methods

File: xbmc/utils/URIUtils.cpp [hunk 12]
```diff
@@ -88,35 +88,80 @@ bool URIUtils::HasExtension(const std::string& strFileName)
   return iPeriod != std::string::npos && strFileName[iPeriod] == '.';
 }
 
-bool URIUtils::HasExtension(const CURL& url, const std::string& strExtensions)
+bool URIUtils::HasExtension(const CURL& url, std::string_view strExtensions)
```
Comment (ksooo) [inline xbmc/utils/URIUtils.cpp:91]: Could you please remove the type prefix from the string param?

=> https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#87-methods

File: xbmc/utils/URIUtils.cpp [hunk 13]
```diff
@@ -88,35 +88,80 @@ bool URIUtils::HasExtension(const std::string& strFileName)
...
 {
   return HasExtension(url.GetFileName(), strExtensions);
 }
 
-bool URIUtils::HasExtension(const std::string& strFileName, const std::string& strExtensions)
+namespace
 {
-  if (IsURL(strFileName))
+bool IsExtensionInList(std::string_view extension, std::string_view extList)
```
Comment (ksooo) [inline xbmc/utils/URIUtils.cpp:98]: Could you please remove the type prefix from the string param?

=> https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#87-methods

File: xbmc/utils/test/TestStringUtils.cpp [hunk 14]
```diff
@@ -134,6 +134,15 @@ TEST(TestStringUtils, EqualsNoCase)
 
   EXPECT_TRUE(StringUtils::EqualsNoCase(refstr, "TeSt"));
   EXPECT_TRUE(StringUtils::EqualsNoCase(refstr, "tEsT"));
+  EXPECT_FALSE(StringUtils::EqualsNoCase(refstr, "tEsTe"));
+  EXPECT_FALSE(StringUtils::EqualsNoCase(refstr, "tEs"));
+  EXPECT_FALSE(StringUtils::EqualsNoCase(refstr, "tEsU"));
+  EXPECT_FALSE(StringUtils::EqualsNoCase(refstr, "ta"));
+  EXPECT_FALSE(StringUtils::EqualsNoCase(refstr, ""));
+
+  std::string empstr = "";
```
Comment (ksooo) [inline xbmc/utils/test/TestStringUtils.cpp:143]: No need for the `= ""`. default ctor is sufficient.

File: xbmc/settings/AdvancedSettings.h (hunk 2 above)
Comment (CrystalP) [reply xbmc/settings/AdvancedSettings.h]: no, I missed it. Added doxygen while I was at it.

File: xbmc/settings/AdvancedSettings.h (hunk 3 above)
Comment (CrystalP) [reply xbmc/settings/AdvancedSettings.h]: Yes, done. Access to a lot of class members should probably be protected but that would be best done as a dedicated PR. Rarely if ever is a problem because settings are typically loaded once on start and there is no uncontrolled activity. Profile switch may be more risky.

File: xbmc/utils/StringUtils.h (hunk 10 above)
Comment (CrystalP) [reply xbmc/utils/StringUtils.h:145]: That's an unintended/unavoidable byproduct of trying to include the choice of 2 algorithms in the PR, one of which needs StringUtils changes, and the other one doesn't (and reverts the prev. StringUtils changes - but reverting to original generates formatting warnings).
If algo 1 / commit 2 is chosen, then the .h and .cpp format changes will match.
If the algorithm 2 / commit 3 is chosen, I'll remove the StringUtils changes+revert and the header won't change at all.

**Comment by CrystalP** [comment]
Jenkins build this please

Thanks for the review, comments addressed or replied to.
Added a few unit tests to the first commit.

File: xbmc/utils/FileExtensionProvider.h (hunk 8 above)
Comment (CrystalP) [reply xbmc/utils/FileExtensionProvider.h:86]: It's initialized in the constructor with the other members.

File: xbmc/utils/FileExtensionProvider.cpp (hunk 7 above)
Comment (CrystalP) [reply xbmc/utils/FileExtensionProvider.cpp:267]: its value doesn't depend on addons.

File: xbmc/utils/FileExtensionProvider.cpp (hunk 6 above)
Comment (CrystalP) [reply xbmc/utils/FileExtensionProvider.cpp:239]: same answer as other similar code - those 2 don't depend on addons.

**Comment by garbear** [comment]
Wanted for Beta 2?
